Add: Disease - Babylonian Fever

timchigaming opened the pull request:
## Описание

### Добавляем НОВЫЙ!! вирус
> Что же делает этот вирус?

1. Заставляет забыть носителя ВСЕ языки на время пока он не вылечится от заразы.
2. Возвращает все языки обратно когда излечивается.
3. Лечится митоколидом + маннитолом

> Но что если кукла изучит новые языки во время работы вируса?

- Он сохранит его в список, очистит список языков куклы и вернёт свежеизученные языки после излечения.

## Ссылка на предложение/Причина создания ПР
https://discord.com/channels/617003227182792704/1256557228807295038/1258174752145539206

## Демонстрация изменений

https://github.com/user-attachments/assets/d3f880f6-2209-438b-ab3f-69421cb90956


## Тесты
Всё работает отлажено, _теперь_.

timchigaming committed:
Готово к мержу.

timchigaming committed:
Merge branch 'ss220-space:master220' into Vavilone-Virusolone

timchigaming committed:
1

timchigaming committed:
Merge branch 'ss220-space:master220' into Vavilone-Virusolone

timchigaming committed:
Рефакторинг и мем 220

timchigaming committed:
Merge branch 'Vavilone-Virusolone' of https://github.com/timchigaming/Paradise into Vavilone-Virusolone

timchigaming committed:
Двести.. Двадцать..

timchigaming committed:
Двести.. Двадцать.. Один..

Den109G commented on code/datums/diseases/viruses/babylon_feve.dm:
```diff
@@ -0,0 +1,64 @@
+					"Уээ... Уэээ... УЭЭЭЭ...",
+					"Ээ... ээ... ээ...",
+					"Ии... ии... ии...",
+					"Двести... двадцать..."
```
Проставь `\` на конце строк, чтобы программа понимала, что это всё один лист.
```suggestion
					"Двести... двадцать..."\
```

Den109G commented on code/datums/diseases/viruses/babylon_feve.dm:
```diff
@@ -0,0 +1,64 @@
+		stored_languages = affected_mob.languages.Copy()
+	else
+		if(affected_mob.languages)
+			stored_languages += affected_mob.languages.Copy()
```
Такая проверка на языки может привести к языкам дубликатам, что засорит чат игрокам.

Den109G commented on code/datums/diseases/viruses/babylon_feve.dm:
```diff
@@ -0,0 +1,64 @@
+			stored_languages += affected_mob.languages.Copy()
+
+	// Remove existing languages
+	affected_mob.languages.Cut()
```
Это надо поменять на `remove_language`

Den109G commented on code/datums/diseases/viruses/babylon_feve.dm:
```diff
@@ -0,0 +1,64 @@
+	if(..())
+		// Restore previously known languages
+		if(stored_languages.len)
+			affected_mob.languages = stored_languages.Copy()
```
А это - на `add_language`.

Den109G reviewed (changes requested):
Выше оставил замечания, что надо менять.

timchigaming committed:
add_language(lan.name)

Den109G commented on code/datums/diseases/viruses/babylon_feve.dm:
```diff
@@ -0,0 +1,67 @@
+				)
+
+	// Store languages on first stage activation
+	if(stage == 1 && !stored_languages.len && affected_mob.languages)
```
Эта проверка особа не нужна. Если она находится вне свитча, потому что болезнь может перейти сразу на вторую стадию, то ставить проверку на первую стадию особо не имеет смысла.

Den109G commented:
Как в #5332 написали, проверять каждую **секунду** что у больного появились языки - not good. Лучше поменять на сигналы, чтобы вирус не крашил кадры в секунду у сервера, если заболело 3 человека.

timchigaming committed:
unif(stage==1)

timchigaming committed:
REGISTER_SIGNAL

timchigaming committed:
NEW

timchigaming committed:
NEW(220)

timchigaming committed:
NEW CURSE(220)

timchigaming committed:
TOTALLY CURSED

timchigaming committed:
FINNALY! CHOKOPAY!

Den109G commented on code/datums/diseases/viruses/babylon_feve.dm:
```diff
@@ -0,0 +1,77 @@
+
+	return TRUE
+
+/datum/disease/virus/babylonian_fever/cure()
```
Надо прописать аргументы у прока.
```suggestion
/datum/disease/virus/babylonian_fever/cure(id = type, need_immunity = TRUE)
```

timchigaming committed:
Not a finnaly chokopay, but arguments for proc

Co-authored-by: Den109G <87372121+Den109G@users.noreply.github.com>

timchigaming committed:
Rrrrrawr

timchigaming committed:
Rawr~

LiquidPotroh commented on code/datums/diseases/viruses/babylon_feve.dm:
```diff
@@ -0,0 +1,78 @@
+	if(!disease)
+		return FALSE
+
+	affected_mob = disease.affected_mob
```
перевожу что у тебя тут написано:
Есть моб Петя, у него вирус А, он заражает моба М, создаём новый экземпляр вируса Б, заражаем им моба М. Меняем у вируса А ссылку на заражённого моба на М, вместо пети. Теперь у моба М целых 2 копии вируса! (если бы не встроенные проверки на дурака разумеется).
Аффектед моб присваивается в ..(). Удали

timchigaming replied:
Да, но... нет, 
![GVELxlvFY1](https://github.com/user-attachments/assets/cf2e9bb7-e191-47a7-956b-000b6b047431)
оно бы жаловалось рантаймом.

LiquidPotroh replied:
Потому что в регистерсигнал надо тоже к дизиз.аффектедмобу привязывать. 
И убери уже двойные скобки в cure(), не видишь, линтер ругается

timchigaming committed:
Finnaly

timchigaming committed:
This is not the end yet, but it is close

timchigaming committed:
блятъ

timchigaming committed:
блят

Den109G commented on code/modules/mining/lavaland/loot/tendril_loot.dm:
```diff
@@ -116,7 +116,13 @@
-/obj/item/book_of_babel/attack_self(mob/user)
+/obj/item/book_of_babel/attack_self(mob/living/carbon/user)
+	if(HAS_TRAIT(user, TRAIT_NO_BABEL))
+		user.visible_message("<span class='notice'>[user] suddenly stops, releasing the [src]</span>")
```
Это поменять на дефайн.
```suggestion
		user.visible_message(span_notice("[user] suddenly stops, releasing the [src]"))
```

Den109G committed:
define span

Daeberdir committed:
Merge branch 'master220' into Babylonian-Fever

Daeberdir commented on code/datums/diseases/viruses/babylon_fever.dm:
```diff
@@ -0,0 +1,77 @@
+
+	LAZYCLEARLIST(stored_languages)
+
+	REMOVE_TRAIT(affected_mob, TRAIT_NO_BABEL, UNIQUE_TRAIT_SOURCE(disease))
```
```suggestion
	REMOVE_TRAIT(affected_mob, TRAIT_NO_BABEL, UNIQUE_TRAIT_SOURCE(src))
```

Daeberdir committed:
Update code/datums/diseases/viruses/babylon_fever.dm

Daeberdir committed:
Maybe Better.

Языки успешно сохраняются и удаляются из списка именно болезни, пока болезнь активна.

Daeberdir commented on code/__DEFINES/mobs.dm:
```diff
@@ -440,6 +440,9 @@
 #define SHOCK_DELAY_STUN (1<<6)
 /// Makes the weaken into a knockdown
 #define SHOCK_KNOCKDOWN (1<<7)
+/// For babylon fever,
+#define DISEASE_LIVING_LANGUAGE_PROCESSED (1<<0)
+
```
```suggestion
```
В твоём случае всё же не нужно, так как альтернативные исходы обработки языков через болезнь не предусмотрены.

Daeberdir committed:
Update code/__DEFINES/mobs.dm

timchigaming committed:
Merge branch 'master220' into Babylonian-Fever

Daeberdir committed:
Done.

Daeberdir reviewed (approved):
Теперь это будет работать.

Den109G reviewed (approved):
Выглядит хорошо со стороны кода.

timchigaming committed:
Merge branch 'master220' into Babylonian-Fever
